### ISSUE TITLE
Add winevdm to upstream Wine so 16-bit apps can run in macOS?

### ISSUE THREAD

**Issue by emendelson** [otya128/winevdm#893]
Probably this is not possible, but I thought I would ask.

Is it possible to add winevdm/otvdm technology to Wine itself, so that 16-bit apps can run in Wine under 64-bit only macOS?

I tried installing otvdm into a Wineskin Winery "wrapper"; the otvdm.exe program ran, but could not run any application. Wineskin Winery is here:

https://github.com/Gcenx/WineskinServer

Again, this may not be possible, and it would presumably require building otvdm.exe as a 64-bit application, which also may not be possible. Please delete this message if it only wastes time.

**Comment by cracyc** [comment]
If wine developers want to take bits that i've written back into wine that's fine but I think you mean to run it on wine not add it to wine itself.  Winevdm (this one rather than the original which is also called winevdm) doesn't run on wine right now because wine tries to load it's own win16 dlls in place of ours.  Since winevdm is a fairly normal win32 program, it runs on reactos which also includes large bits of wine but has it's own ntvdm, it could work if a way to keep it from getting tangled with wine's own winevdm.

**Comment by otya128** [comment]
Anyway, in the 32-bit version of wine, you can execute it by overriding the library and giving priority to native such as krnl386.exe16.
https://wiki.winehq.org/Winecfg#Libraries

It's not impossible to port to 64-bit (especially if /LARGEADDRESSAWARE:NO), but there are issues such as having to compile and porting some thunk code.

**Comment by emendelson** [comment]
Thank you for the clarifications. For running in macOS, it seems that everything would need to be ported to 64-bit, but that's obviously a very large project. But of course I don't know what I'm talking about; I'm only guessing. 

Still, something like this (either running it in wine, or adding it to wine somehow) seems to be the only practical hope for running 16-bit Windows applications in the current version of macOS.

**Comment by cracyc** [comment]
> For running in macOS, it seems that everything would need to be ported to 64-bit because 32-bit code won't run in current macOS,

Yes, although Crossover supports 32bit code on catalina and big sur (I had them conflated in my mind when I wrote my reply).

**Comment by romanbtl** [comment]
Running otvdm with Wine should be possible with the soon to be released version 6.11 of Wine, see:
https://bugs.winehq.org/show_bug.cgi?id=51224

It would be great if a wrapper that adjusts Wine overrides properly could be packed with otvdm binaries.

**Comment by cracyc** [comment]
KVM support would be neat too.  I think that GVM is closely related to KVM which is already supported.

**Comment by romanbtl** [comment]
If that was not clear, my suggestion was about including just one file (something like https://bugs.winehq.org/attachment.cgi?id=70122) or updating readme, so Mac and Linux users have easier life and don't spend time investigating how dll overrides work and how to set them. This suggestion and confirmation that it works on Linux is all how I can contribute. Thanks.

**Comment by stuaxo** [comment]
@romanbtl it sounds like updating the README would be a good start.

I would start from a fresh wine install and make notes as you go as to what you need to do, then you can edit the README here and make a PR.
